plugins/ltac2/tac2intern.ml

```diff
@@ -213,38 +213,42 @@ let rec intern_type env ({loc;v=t} : raw_typexpr) : UF.elt glb_typexpr = match t
 | CTypVar (Name id) -> GTypVar (get_alias (CAst.make ?loc id) env)
 | CTypVar Anonymous -> GTypVar (fresh_id env)
 | CTypRef (rel, args) ->
-  let (kn, nparams) = match rel with
+  let (kn, knloc, nparams) = match rel with
   | RelId qid ->
     let id = qualid_basename qid in
     if qualid_is_ident qid && Id.Map.mem id env.env_rec then
       let (kn, n) = Id.Map.find id env.env_rec in
-      (Other kn, n)
+      (Other kn, qid.CAst.loc, n)
     else
       let kn =
         try Tac2env.locate_type qid
         with Not_found ->
           user_err ?loc (str "Unbound type constructor " ++ pr_qualid qid)
       in
       let (nparams, _) = Tac2env.interp_type kn in
-      (Other kn, nparams)
-  | AbsKn (Other kn) ->
+      (Other kn, qid.CAst.loc, nparams)
+  | AbsKn (Other {loc;v=kn}) ->
     let (nparams, _) = Tac2env.interp_type kn in
-    (Other kn, nparams)
+    (Other kn, loc, nparams)
   | AbsKn (Tuple n) ->
-    (Tuple n, n)
+    (Tuple n, loc, n)
   in
   let nargs = List.length args in
   let () =
     if not (Int.equal nparams nargs) then
       let qid = match rel with
       | RelId lid -> lid
-      | AbsKn (Other kn) -> shortest_qualid_of_type ?loc kn
+      | AbsKn (Other {loc;v=kn}) -> shortest_qualid_of_type ?loc kn
       | AbsKn (Tuple _) -> assert false
       in
       user_err ?loc (strbrk "The type constructor " ++ pr_qualid qid ++
         strbrk " expects " ++ int nparams ++ strbrk " argument(s), but is here \
         applied to " ++ int nargs ++ strbrk "argument(s)")
   in
+  let () =
+    match kn with
+    | Other kn -> Dumpglob.dump_knref ?loc:knloc kn "tac2type"
```

Comment:
The dumpglob mechanism is fairly broken, I recently faced a conundrum while trying to get rid of kernel side-effects because it decided that a kernel name was missing in the global env (while existing locally) and died horribly. I'd rather have an Ltac2 implementation that is future-proof from scratch than having to port it later.

Reply:
Then I am closing this pull request because I do not have the time to redesign Dumpglob from scratch.

Reply:
What? It doesn't change anything from the point of view of dumpglob. I'll reopen this and add a patch atop of it, but it will need a distinct assignee.

Reply:
I can assign if needed.

Reply:
> What? It doesn't change anything from the point of view of dumpglob.

It changes everything. Ltac2 terms embed Gallina terms and so on. As such, file `tac2intern.ml` invokes `Genintern.intern`, which in turn can invoke an internalization function from about anywhere in Coq, not just the Ltac2 plugin. If you want a new Dumpglob mechanism with delayed semantics for Ltac2, you also need it for all the parts currently accessible from `Genintern`. We are talking about a major rewrite here.